### ISSUE TITLE
Error while running migration version 3 for 'OrchardCore.Lists'

### ISSUE THREAD

**Issue by DrewBrasher** [OrchardCMS/OrchardCore#13043]
Getting error after upgrading from OrchardCore 1.4 to 1.5 when connected to MS SQL Server 2012. It works fine in development using SQL Lite but when I publish the the live server which uses MS SQL 2012 I get the following error:

```
Default|00-b3adf13172631d88faf6e253ae24016b-d9a025b508204208-00||OrchardCore.Data.Migration.DataMigrationManager|ERROR|Error while running migration version 3 for 'OrchardCore.Lists'. System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
---> Microsoft.Data.SqlClient.SqlException (0x80131904): Incorrect syntax near the keyword 'if'.
at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
at Microsoft.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds(String methodName, Boolean isAsync, Int32 timeout, Boolean asyncWrite)
at Microsoft.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, Boolean sendToPipe, Int32 timeout, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry, String methodName)
at Microsoft.Data.SqlClient.SqlCommand.ExecuteNonQuery()
at Dapper.SqlMapper.ExecuteCommand(IDbConnection cnn, CommandDefinition& command, Action`2 paramReader) in /_/Dapper/SqlMapper.cs:line 2848
at Dapper.SqlMapper.ExecuteImpl(IDbConnection cnn, CommandDefinition& command) in /_/Dapper/SqlMapper.cs:line 581
at YesSql.Sql.SchemaBuilder.Execute(IEnumerable`1 statements)
at YesSql.Sql.SchemaBuilder.AlterTable(String name, Action`1 table)
at YesSql.Sql.SchemaBuilder.AlterIndexTable(Type indexType, Action`1 table, String collection)
at YesSql.Sql.SchemaBuilderExtensions.AlterIndexTable[T](ISchemaBuilder builder, Action`1 table, String collection)
at OrchardCore.Lists.Migrations.UpdateFrom3()
ClientConnectionId:65f68ead-5665-4adf-99a9-28181cb5e3e4
Error Number:156,State:1,Class:15
--- End of inner exception stack trace ---
at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
at OrchardCore.Data.Migration.DataMigrationManager.UpdateAsync(String featureId) at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
at OrchardCore.Data.Migration.DataMigrationManager.UpdateAsync(String featureId) 
```

**Comment by Skrypt** [comment]
Using SQL Server 2012 may always cause issues like these. The SQL syntax is different from newer versions. But it is the first time I see an issue with AlterIndexTable. In that case, it is more related to YesSQL and what it supports.

**Comment by DrewBrasher** [comment]
Is there a way I could see the SQL generated by the migration? Maybe I could modify the sql to work with 2012 and run it manually. Or is my best option to stay on 1.4 until we upgrade our SQL Server?

**Comment by Skrypt** [comment]
Your best bet would be to do a SQL schema compare between a database that has these index tables. Just create a new tenant from a clean 1.5 source code solution and then compare it with the database that failed this migration.

**Comment by DrewBrasher** [comment]
Thanks for your help @Skrypt, I did what you suggested and the site is working but it still logs an exception on startup because it is still trying to run that migration. Is the last migration number run stored somewhere that I can edit?

**Comment by Skrypt** [comment]
It is inside the Document table.

![image](https://user-images.githubusercontent.com/3228637/210852137-97f24790-c3ba-450a-857a-98f9169f4c53.png)

You willl need to set the migration number to 4.

**Comment by sebastienros** [comment]
I don't want to have older sql dialects (2012) because there is no way to test them and ensure it works. And also we would need to set the dialect based on another configuration. If you follow @Skrypt 's advice you should be able to force the version of the migration to skip it.

**Comment by DrewBrasher** [comment]
@Skrypt Thanks again for all your help! I modified the document like you said and it is working without any errors now.
In case anyone else runs across this, the SQL I had to run to add the columns that the migration does was this:
```
ALTER TABLE ContainedPartIndex
ADD 
	ContentItemId varchar(26) null,
	DisplayText varchar(255) null,
	ListContentType varchar(255) null,
	Published bit null,
	Latest varchar(255) null;
```
@sebastienros I totally understand not supporting SQL 2012. Hopefully we will upgrade soon. This issue can be closed.
